src/interpolation_functions.jl

```diff
@@ -0,0 +1,304 @@
+  end
+  while last_col > 0 && all(matrix[:, last_col] .== 0)
+    last_col -= 1
+  end
+  return matrix[1:last_row, 1:last_col]
+end
+
+""" 
+    function_itensornetwork(s,f, cutoff=1e-3, max_coeffs=100, mode="fourier", by_mag=true)
+    Takes in a function `f` as a black box, outputs a tensor network approximation using Fourier or Chebyshev series. Supports 1D or 2D functions.
+
+    #Arguments
+    - `s`: IndsNetworkMap
+    - `f`: a real-valued function on domain [0,1] or [0,1]⊗[0,1] which can be given as a black box
+    - `cutoff`: (default=1e-3) specifies at what value should the fourier/chebyshev coefficients be truncated below
+    - `max_coeffs`: the maximum number of coefficients allowed in 1D. In 2D, it is the maximum degree of a coefficient in each dimension
+    - `mode`: ["fourier", "chebyshev"] "fourier" uses a fourier basis and "chebyshev" uses a chebyshev polynomial basis
+    - `by_mag`: boolean. if false, will use the `max_coeffs` coefficients of lowest degree. if true, will use the `max_coeffs` coefficients of greatest absolute magnitude.
+    - `trunc_level`: what cutoff the periodic truncation is done at
+"""
+function function_itensornetwork(
+  s::IndsNetworkMap,
+  f::Function;
+  cutoff::Float64=1e-3,
+  max_coeffs::Integer=100,
+  mode="fourier",
+  by_mag=false,
+  trunc_level=1e-16,
+)
+
+  #get the number of inputs to f
+  method_list = methods(f).ms
+  arg_counts = [length(m.sig.parameters) - 1 for m in method_list]
+  num_inputs = arg_counts[1]
+
+  """ 1D functions """
+  if num_inputs == 1
+    if mode == "fourier"
+      S = Fourier(Interval(0.0, 1.0))
+      f_fourier = Fun(f, S)
+      a = chop(f_fourier, cutoff)
+      if by_mag
+        cf = greatest_n(coefficients(a), max_coeffs)
+      else
+        cf = ncoefficients(a) > max_coeffs ? coefficients(a)[1:max_coeffs] : coefficients(a)
+      end
+      ψ = fourier_itensornetwork(s, cf; cutoff=trunc_level)
+    elseif mode == "chebyshev"
+      T = Chebyshev(Interval(0.0, 1.0))
+      f_chebyshev = Fun(f, T)
+      a = chop(f_chebyshev, cutoff)
+      if by_mag
+        cf = greatest_n(coefficients(a), max_coeffs)
+      else
+        cf = ncoefficients(a) > max_coeffs ? coefficients(a)[1:max_coeffs] : coefficients(a)
+      end
+      ψ = chebyshev_itensornetwork(s, cf; cutoff=trunc_level)
+    else
+      throw("mode $mode not recognized")
+    end
+    return ψ
+
+    """ 2D functions """
+  elseif num_inputs == 2
+    if mode == "fourier"
+      S = Fourier(Interval(0.0, 1.0))^2
+      f_fourier = ProductFun(f, S; tol=cutoff)
+      if by_mag
+        cf = greatest_n_2d(coefficients(f_fourier), max_coeffs)
+      else
+        a = min(max_coeffs, size(coefficients(f_fourier))[1])
+        b = min(max_coeffs, size(coefficients(f_fourier))[2])
+        cf = coefficients(f_fourier)[1:a, 1:b]
+      end
+      ψ = fourier_2D_itensornetwork(s, cf; dims=[1, 2], cutoff=trunc_level)
+    elseif mode == "chebyshev"
+      T = Chebyshev(Interval(0.0, 1.0))^2
```

Comment:
It seems we could just do `Chebyshev(Interval(0.0, 1.0))^num_inputs` rather than forcing only 2, is there anything stopping us from generalizing?

Reply:
The main reason it's not generalized to higher dimensions is that the ApproxFun library can only do interpolation in 1D and 2D. [Chebfun3](https://www.chebfun.org/docs/guide/guide18.html) can do chebyshev/fourier interpolation in up to 3 dimensions, but afaik there's no equivalent julia library, and it seems like in general doing function interpolation in higher dimensions is extremely non-trivial.

Reply:
Makes sense, lets not do that yet then